doc/lsst.pipe.base/command-line-task-parallel-howto.rst

```diff
@@ -0,0 +1,97 @@
+
+.. note::
+
+   Internally, :option:`-j`-type multiprocessing is implemented with the `multiprocessing` module in the Python standard library.
+
+   If a task does not support :option:`-j` parallel processing it will report a non-fatal warning:
+
+   .. code-block:: text
+
+      This task does not support multiprocessing; using one process
+
+.. _command-line-task-parallel-howto-distributed:
+
+How to enable distributed processing
+====================================
+
+Command-line tasks implemented in the ``pipe_drivers`` package can distribute processing across multiple nodes using MPI or batch submission to clusters (PBS or SLURM).
+See the ``pipe_drivers`` documentation for more information.
+
+.. _command-line-task-parallel-howto-threading:
+
+How to control threading while parallel processing
+==================================================
+
+LSST Science Pipelines uses low-level math libraries that support threading.
+OpenBLAS_ and MKL_ default to using the same number of threads as CPUs present on the machine.
+When parallel processing at a higher level (using :option:`-j`, for example), threading in these libraries can result in a **net decrease in performance** because of thread contention.
+
+When running with :option:`-j` multiprocessing or via ``ctrl_pool``, command-line tasks check if low-level libraries are using their default threading behavior (same number of threads as CPUs).
+If so, the task will automatically disable multi-threading in these libraries.
+The command-line task also sends a warning when it does so:
+
+.. code-block:: text
+
+   WARNING: You are using OpenBLAS with multiple threads (16), but have not specified the number of threads using one of the OpenBLAS environment variables: OPENBLAS_NUM_THREADS, GOTO_NUM_THREADS, OMP_NUM_THREADS. This may indicate that you are unintentionally using multiple threads, which may cause problems. WE HAVE THEREFORE DISABLED OpenBLAS THREADING. If you know what you are doing and want threads enabled implicitly, set the environment variable LSST_ALLOW_IMPLICIT_THREADS.
+
+The recommended resolution to this warning is to specifically limit the number of threads used by OpenBLAS_ and MKL_.
+For example, in a :command:`bash` or similar shell set the ``OMP_NUM_THREADS`` environment variable:
+
+.. code-block:: bash
+
+   export OMP_NUM_THREADS=1
+
+.. note::
+
+   ``OMP_NUM_THREADS`` is recognized by both OpenBLAS_ and MKL_ so it is likely the only environment variable that needs to be set.
+
+   Alternatively, specific environment variables used by the libraries are:
+
+   - OpenBLAS: ``OPENBLAS_NUM_THREADS``, ``GOTO_NUM_THREADS``, and ``OMP_NUM_THREADS``.
+   - MLK: ``MKL_NUM_THREADS``, ``MKL_DOMAIN_NUM_THREADS``, ``OMP_NUM_THREADS``.
+
+If necessary, you may bypass a command-line task's control over threading in low-level math libraries by setting the ``LSST_ALLOW_IMPLICIT_THREADS`` environment variable:
+
+.. code-block:: bash
+
+   export LSST_ALLOW_IMPLICIT_THREADS=1
+
```

Comment:
While the preferred unix shell for scripting has been Bourne shell since unix began, people do use csh.  You should probably have a note on csh environment variable syntax somewhere.